src/MongoDB.Driver.Core/Core/Bindings/CoreServerSessionPool.cs

```diff
@@ -36,24 +34,17 @@ public CoreServerSessionPool(ICluster cluster)
         /// <inheritdoc />
         public ICoreServerSession AcquireSession()
         {
-            lock (_lock)
+            ICoreServerSession pooledSession;
+            while (_pool.TryAcquire(out pooledSession))
             {
-                for (var i = _pool.Count - 1; i >= 0; i--)
+                if (IsAboutToExpire(pooledSession))
                 {
-                    var pooledSession = _pool[i];
-                    if (IsAboutToExpire(pooledSession))
-                    {
-                        pooledSession.Dispose();
-                    }
-                    else
-                    {
-                        var removeCount = _pool.Count - i; // the one we're about to return and any about to expire ones we skipped over
-                        _pool.RemoveRange(i, removeCount);
-                        return new ReleaseOnDisposeCoreServerSession(pooledSession, this);
-                    }
+                    pooledSession.Dispose();
```

Comment:
A key element of this change is that we are now calling Dispose *outside* the lock.